pgjdbc/src/main/java/org/postgresql/core/ParameterContext.java

```diff
@@ -0,0 +1,166 @@
+
+public class ParameterContext {
+  public static final ParameterContext EMPTY_CONTEXT = new ParameterContext();
+  private static final List<Integer> NO_PLACEHOLDERS = new ArrayList<>();
+  private @Nullable BindStyle bindStyle = null;
+  private @Nullable List<Integer> placeholderPositions = null;
+  private @Nullable List<String> placeholderNames = null;
+  private @Nullable List<Integer> placeholderAtPosition = null;
+
+  public static ParameterContext buildNamed(List<Integer> placeholderPositions,
+      List<String> placeholderNames) throws SQLException {
+    final ParameterContext ctx = new ParameterContext();
+    assert placeholderPositions.size() == placeholderNames.size();
+    for (int i = 0; i < placeholderPositions.size(); i++) {
+      ctx.addNamedParameter(placeholderPositions.get(i), placeholderNames.get(i));
+    }
+
+    return ctx;
+  }
+
+  public int addPositionalParameter(int position) throws SQLException {
+    if (bindStyle == BindStyle.NAMED) {
+      throw new SQLException("Positional and named parameters cannot be combined! Saw named"
+          + " parameter first.");
+    }
+
+    if (bindStyle == null) {
+      bindStyle = BindStyle.POSITIONAL;
+    }
+
+    if (placeholderPositions == null) {
+      placeholderPositions = new ArrayList<>();
+    }
+
+    placeholderPositions.add(position);
+    int bindIndex = placeholderPositions.size() - 1;
+
+    if (placeholderAtPosition == null) {
+      placeholderAtPosition = new ArrayList<>();
+    }
+
+    placeholderAtPosition.add(bindIndex);
+
+    return bindIndex + 1;
+  }
+
+  public boolean hasParameters() {
+    return placeholderPositions != null && !placeholderPositions.isEmpty();
+  }
+
+  public boolean hasNamedParameters() {
+    return placeholderNames != null && !placeholderNames.isEmpty();
+  }
+
+  /**
+   * @param i 0-indexed position in the order of first appearance
+   * @return The name of the placeholder at this backend parameter position
+   */
+  public String getPlaceholderName(int i) {
+    if (placeholderNames == null) {
+      throw new RuntimeException("Call hasNamedParameters() first.");
+    }
+    return placeholderNames.get(i);
+  }
+
+  /**
+   * @param i 0-indexed position in the order of appearance
+   * @return The position of the placeholder in the SQL text for this placeholder index
+   */
+  public int getPlaceholderPosition(int i) {
+    if (placeholderPositions == null) {
+      throw new RuntimeException("Call hasParameters() first.");
+    }
+    return placeholderPositions.get(i);
+  }
+
+  /**
+   * @param i 0-indexed position in the order of appearance
+   * @return The position of the placeholder in the order of first appearance of each placeholder
+   */
+  public int getPlaceholderAtPosition(int i) {
+    if (placeholderAtPosition == null) {
+      throw new RuntimeException("Call hasParameters() first.");
+    }
+    return placeholderAtPosition.get(i);
+  }
+
+  public int getLatestPlaceholderPosition() {
```

Comment:
would this be better named `getLastPlaceholderPosition`

Reply:
Fine with me.